rfcs/0007-filtering.md

```diff
@@ -0,0 +1,211 @@
+* Start Date: 2020-08-21
+* RFC PR: [rfcs#0007](https://github.com/CMSgov/beneficiary-fhir-data/pull/345)
+* JIRA Ticket(s):
+    * [https://jira.cms.gov/browse/AB2D-1863](https://jira.cms.gov/browse/AB2D-1863)
+
+This proposal will suggest adding server side filtering to API calls that return EOBs to the client in order to 
+reduce on the amount of data that needs to be transferred. 
+
+## Table of Contents
+[Table of Contents]: #table-of-contents
+
+* [RFC Proposal](#rfc-proposal)
+* [Table of Contents](#table-of-contents)
+* [Motivation](#motivation)
+* [Proposed Solution](#proposed-solution)
+    * [Proposed Solution: Detailed Design](#proposed-solution-detailed-design)
+    * [Proposed Solution: Unresolved Questions](#proposed-solution-unresolved-questions)
+    * [Proposed Solution: Drawbacks](#proposed-solution-drawbacks)
+    * [Proposed Solution: Notable Alternatives](#proposed-solution-notable-alternatives)
+* [Future Possibilities](#future-possibilities)
+
+## Motivation
+[Motivation]: #motivation
+
+AB2D receives more data than needed from BFD when we make API calls to gather EOBs (Explanation of Benefits).  AB2D
+currently filters out unrelevant fields.  This might not be terrible for dozens of API calls, but for 
+10 to 100 thousand requests it can impact our performance significantly.  Since we only need a projection of the data
+(10 out of 69 top level getters; 6 out of 42 fields in the ItemComponent), it makes
+sense for the server to send us only the data consumed.  This reduces bandwidth requirements for receiving the data.
+
+For example, GraphQL solves this problem by allowing clients to specify field selection (e.g. a database projection) at
+invocation time.
+
+## Proposed Solution
+[Proposed Solution]: #proposed-solution
+
+The solution adds code in the BFD server to configure which fields AB2D requires.  Non-required fields are null'd out
+before being transformed in the result.  The specific set of required fields should remain stable
+and not require modifications.
+
+The filtering will need to happen towards the end of processing. This is to avoid
+spaghetti code since transformations from the entity object to the FHIR object happen in several places throughout
+the codebase. 
+
+
+### Proposed Solution: Detailed Design
+[Proposed Solution: Detailed Design]: #proposed-solution-detailed-design
+
+The BeanUtils library can be utilized for transforming properties
+
+```
+<dependency>
+    <groupId>commons-beanutils</groupId>
+    <artifactId>commons-beanutils</artifactId>
+    <version>1.9.4</version>
+</dependency>
+```
+
+A `private static final Map<String, SubfieldSelection>` of fields will need to be maintained in `TransformerUtils`. This will
+be the AB2D required fields in the FHIR object that gets sent back to the client. Just before the
+end of the method `findByPatient` in `ExplanationOfBenefitsResourceProvider` the `eobs` can be passed to a new
+method in `TransformerUtils`.
```

Comment:
For what it's worth, this really doesn't relate to transformation and I'd be more inclined to include it in a new class, e.g. `gov.cms.bfd.server.war.commons.ResourceFieldFilter`.